lib/shim/message-shim.js

```diff
@@ -356,108 +354,37 @@ function recordConsume(nodule, properties, spec) {
-  return this.wrap(nodule, properties, function wrapConsume(shim, fn, fnName) {
-    if (!shim.isFunction(fn)) {
-      shim.logger.debug('Not wrapping %s (%s) as consume', fn, fnName)
-      return fn
-    }
-
-    return function consumeRecorder() {
-      const parent = shim.getSegment()
-      if (!parent || !parent.transaction.isActive()) {
-        shim.logger.trace('Not recording consume, no active transaction.')
-        return fn.apply(this, arguments)
-      }
-
-      // Process the message args.
-      const args = shim.argsToArray.apply(shim, arguments)
-      let msgDesc = null
-      if (shim.isFunction(spec)) {
-        msgDesc = spec.call(this, shim, fn, fnName, args)
-        shim.setDefaults(msgDesc, DEFAULT_SPEC)
-      } else {
-        msgDesc = {
-          destinationName: null,
-          callback: spec.callback,
-          promise: spec.promise,
-          messageHandler: spec.messageHandler
-        }
-
-        const destIdx = shim.normalizeIndex(args.length, spec.destinationName)
-        if (destIdx !== null) {
-          msgDesc.destinationName = args[destIdx]
-        }
-      }
-
-      // Make the segment if we can.
-      if (!msgDesc) {
-        shim.logger.trace('Not recording consume, no message descriptor.')
-        return fn.apply(this, args)
-      }
-
-      const name = _nameMessageSegment(shim, msgDesc, shim._metrics.CONSUME)
-
-      // Adds details needed by createSegment when used with a spec
-      msgDesc.name = name
-      msgDesc.recorder = genericRecorder
-      msgDesc.parent = parent
-
-      const segment = shim.createSegment(msgDesc)
-      const getParams = shim.agent.config.message_tracer.segment_parameters.enabled
-      const resHandler = shim.isFunction(msgDesc.messageHandler) ? msgDesc.messageHandler : null
-
-      const cbIdx = shim.normalizeIndex(args.length, msgDesc.callback)
-      if (cbIdx !== null) {
-        shim.bindCallbackSegment(args, cbIdx, segment)
-
-        // If we have a callback and a results handler, then wrap the callback so
-        // we can call the results handler and get the message properties.
-        if (resHandler) {
-          shim.wrap(args, cbIdx, function wrapCb(shim, cb, cbName) {
-            if (shim.isFunction(cb)) {
-              return function cbWrapper() {
-                const cbArgs = shim.argsToArray.apply(shim, arguments)
-                const msgProps = resHandler.call(this, shim, cb, cbName, cbArgs)
-                if (getParams && msgProps && msgProps.parameters) {
-                  shim.copySegmentParameters(segment, msgProps.parameters)
-                }
-
-                return cb.apply(this, arguments)
-              }
-            }
-          })
-        }
+  return this.record(nodule, properties, function consumeRecorder(shim, fn, fnName, args) {
+    if (shim.isFunction(spec)) {
+      spec = spec.apply(this, arguments)
+    } else {
+      spec = { ...DEFAULT_SPEC, ...spec }
```

Comment:
what if you just change `shim.setDefaults` to do this?  that way it'll encapsulated in API

Reply:
I kind of deliberately didn't use `shim.setDefaults`, because I thought it was rather pointless function. It predates the spread operator. It does make a slight difference because it starts from `Object.create(null)` though. Do you really think it's worth to go back to it?

Reply:
I'm suggesting you change shim.setDefaults to spread to encapsulate this in a function and not just one of spreading in some places and not in others

Reply:
Okay, will do.